src/cache_padded.rs

```diff
@@ -1,120 +1,84 @@
+    _value: T,
+}
+
+/// Pads `T` to the length of a cache line.
 ///
-/// At the moment, cache lines are assumed to be 32 * sizeof(usize) on all
-/// architectures.
+/// Sometimes concurrent programming requires a piece of data to be padded out to the size of a
+/// cacheline to avoid "false sharing": cache lines being invalidated due to unrelated concurrent
+/// activity. Use this type when you want to *avoid* cache locality.
 ///
-/// **Warning**: the wrapped data is never dropped; move out using `ptr::read`
-/// if you need to run dtors.
+/// At the moment, cache lines are assumed to be 64 bytes on all architectures.
+/// Note that, while the size of `CachePadded<T>` is 64 bytes, the alignment may not be.
+#[allow(unions_with_drop_fields)]
 pub struct CachePadded<T> {
-    data: UnsafeCell<[usize; CACHE_LINE]>,
-    _marker: ([Padding; 0], marker::PhantomData<T>),
-}
+    inner: Inner<T>,
 
-impl<T> fmt::Debug for CachePadded<T> {
-    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
-        write!(f, "CachePadded {{ ... }}")
-    }
+    /// `[T; 0]` ensures correct alignment.
+    /// `PhantomData<T>` signals that `CachePadded<T>` contains a `T`.
+    _marker: ([T; 0], PhantomData<T>),
 }
 
 unsafe impl<T: Send> Send for CachePadded<T> {}
 unsafe impl<T: Sync> Sync for CachePadded<T> {}
 
-/// Types for which `mem::zeroed()` is safe.
-///
-/// If a type `T: ZerosValid`, then a sequence of zeros the size of `T` must be
-/// a valid member of the type `T`.
-pub unsafe trait ZerosValid {}
-
-#[cfg(feature = "nightly")]
-unsafe impl ZerosValid for .. {}
-
-macro_rules! zeros_valid { ($( $T:ty )*) => ($(
-    unsafe impl ZerosValid for $T {}
-)*)}
-
-zeros_valid!(u8 u16 u32 u64 usize);
-zeros_valid!(i8 i16 i32 i64 isize);
-
-unsafe impl ZerosValid for ::std::sync::atomic::AtomicUsize {}
-unsafe impl<T> ZerosValid for ::std::sync::atomic::AtomicPtr<T> {}
-
-impl<T: ZerosValid> CachePadded<T> {
-    /// A const fn equivalent to mem::zeroed().
-    #[cfg(not(feature = "nightly"))]
-    pub fn zeroed() -> CachePadded<T> {
-        CachePadded {
-            data: UnsafeCell::new([0; CACHE_LINE]),
-            _marker: ([], marker::PhantomData),
-        }
-    }
+impl<T> CachePadded<T> {
+    /// Pads a value to the length of a cache line.
+    pub fn new(t: T) -> CachePadded<T> {
+        assert!(mem::size_of::<T>() <= mem::size_of::<CachePadded<T>>());
+        assert!(mem::align_of::<T>() <= mem::align_of::<CachePadded<T>>());
 
-    /// A const fn equivalent to mem::zeroed().
-    #[cfg(feature = "nightly")]
-    pub const fn zeroed() -> CachePadded<T> {
-        CachePadded {
-            data: UnsafeCell::new([0; CACHE_LINE]),
-            _marker: ([], marker::PhantomData),
+        unsafe {
+            let mut padded = CachePadded {
+                inner: mem::uninitialized(),
+                _marker: ([], PhantomData),
+            };
+            let p: *mut T = &mut *padded;
+            ptr::write(p, t);
```

Comment:
I want to make sure that this write conforms to C's union semantics.  (I'm assuming Rust's union is the same with C's union.)  Unfortunately I don't know the exact definition in the standard C: I just know that there is a very complex rule that determines what is safe and what is unsafe use of unions.  A good reference would be [Krebber's thesis](http://robbertkrebbers.nl/thesis.html).

Even if it turned out to be inconformant to C, it's probably okay, because for now we only need to think of LLVM semantics.  We don't need to think of the rules on unions when we turn off the `-tbaa` option (type-based alias analysis).